common/lib/xmodule/xmodule/js/src/video/09_completion.js

```diff
@@ -0,0 +1,142 @@
+                this.state.el.off('destroy', this.destroy);
+                delete this.state.completionHandler;
+            },
+
+        /** Initializes the module. */
+            initialize: function() {
+                this.lastSentAt = undefined;
+                this.complete = false;
+                this.completionPercentage = this.state.config.completionPercentage;
+                this.startTime = this.state.config.startTime;
+                this.endTime = this.state.config.endTime;
+                this.isEnabled = this.state.config.completionEnabled;
+                if (this.endTime) {
+                    this.completeAfter = this.calculateCompleteAfter(this.startTime, this.endTime);
+                }
+                if (this.isEnabled) {
+                    this.bindHandlers();
+                }
+            },
+
+        /** Bind any necessary function callbacks to DOM events. */
+            bindHandlers: function() {
+                var self = this;
+
+            /** Event handler to check if the video is complete, and submit a completion if it is.
+             *
+             *  If the timeupdate handler doesn't fire after the required percentage, this will catch
+             *  any fully complete videos.
+             **/
+                this.state.el.on('ended.completion', function() {
+                    self.handleEnded();
+                });
+
+            /** Event handler to check video progress, and mark complete if
+             *  greater than completionPercentage
+             **/
+                this.state.el.on('timeupdate.completion', function(ev, currentTime) {
+                    self.handleCompletion(currentTime);
+                });
+            },
+
+            handleEnded: function() {
+                if (this.complete) {
+                    return;
+                }
+                this.markCompletion();
+            },
+
+            handleCompletion: function(currentTime) {
+                var duration;
+                if (this.complete) {
+                    return;
+                }
+                if (this.lastSentAt !== undefined && currentTime - this.lastSentAt < this.repostDelay()) {
+                    // Throttle attempts to submit in case of network issues
+                    return;
+                }
+                if (this.completeAfter === undefined) {
+                    // Duration is not available at initialization time
+                    duration = this.state.videoPlayer.duration();
+                    if (!duration) {
+                        // duration is not yet set. Wait for another event,
+                        // or fall back to 'ended' handler.
+                        return;
+                    }
+                    this.completeAfter = this.calculateCompleteAfter(this.startTime, duration);
+                }
+
+                if (currentTime > this.completeAfter) {
+                    this.lastSentAt = currentTime;
```

Comment:
> Seems like lastSentAt would be set in markCompletion

I can be down with that.  I'll make the update.

> On a related note, would handleEnded need the same throttling, or is that a single event?

handleEnded is just a single event, so I thought it would be better to keep it simle.

> If handleEnded ajax call fails is the event repeated?

No.  That could be useful, but it brings up a few new questions.  What if the server's down for five minutes?  Or if the user is on a train, and the wifi went out for half an hour?  How often and how long should we repeat?  Should we use local storage to keep stale completions around until the user is back online?  How do we determine when they come back online?  

Ultimately, I think the 100% solution is going to be a much bigger one, and I'm not sure we need to hit 100% (because as you suggest, the user can go back and recomplete it fairly easily).  The question is how good is good enough?  Is a single submission sufficient?  Should we retry three times, every five seconds, and call it quits after that?

> If someone fasts forwards to the end of a video, does that make it seem like it was watched?
Basically, just want to make sure someone can easily mark it as complete if there is a temporary problem.

Yes.  As long as the user is watching a portion of video after the 95% mark when timeupdate gets triggered, the video is considered complete.